Make CPU idleness profiling as an option

derekbit opened the pull request:
The CPU idleness introduces extra overhead and storage performance drop.
By default, the profiling is disabled.

Signed-off-by: Derek Su <derek.su@suse.com>

yasker commented on deploy/fio.yaml:
```diff
@@ -34,6 +34,8 @@ spec:
           value: "/volume/test"
         - name: SIZE
           value: "30G" # must be 10% smaller than the PVC size due to filesystem also took space
+        - name: CPU_IDLE_PROF
```
Thanks for identifying the overhead! I don't think we should keep the option then. We need to think about another way to get the user aware of the potential CPU impact on the latency though.

derekbit replied:
Agree. We need have a way to get average CPU usage by parallel.

yasker commented on README.md:
```diff
@@ -68,14 +74,14 @@ For official benchmarking:
     * It's a measurement of how efficient is the storage system to handle each request.
     * The data path overhead of a storage system can be expressed as the latency it added on top of the native storage system (SSD/NVMe).
-* **CPU Idleness**: How idle is the CPU on the node that's running the test. *Higher is better.*
+* **CPU Utilization**: How busy is the CPU spent in kernel (sys) and in user-mode codes (usr) on the node that's running the test. *Lower is better.*
```
It's not `lower is better`.

https://fio.readthedocs.io/en/latest/fio_doc.html#interpreting-the-output
> cpu
CPU usage. User and system time, along with the number of context switches this thread went through, usage of system and user time, and finally the number of major and minor page faults. The CPU utilization numbers are averages for the jobs in that reporting group, while the context and fault counters are summed.


So if the CPU is lower, it only means less CPU used by FIO, which can be caused by:
1. Other processes took more CPU cycles so there isn't much left for FIO.
2. The other processes cannot use more CPU before hitting their limit so FIO can use more.

I used idleness profile (https://fio.readthedocs.io/en/latest/fio_doc.html#cpu-idleness-profiling) before it's basically taking all the remaining CPU with a lowest priority task, which means those are the CPU that can be used in the environment. So we can say for more if it's higher, the CPU is more `idle`.

derekbit replied:
> It's not `lower is better`.
> 
> https://fio.readthedocs.io/en/latest/fio_doc.html#interpreting-the-output
> 
> > cpu
> > CPU usage. User and system time, along with the number of context switches this thread went through, usage of system and user time, and finally the number of major and minor page faults. The CPU utilization numbers are averages for the jobs in that reporting group, while the context and fault counters are summed.
> 
> So if the CPU is lower, it only means less CPU used by FIO, which can be caused by:
> 
> 1. Other processes took more CPU cycles so there isn't much left for FIO.
> 2. The other processes cannot use more CPU before hitting their limit so FIO can use more.
Right. The statement, "lower is better", is true when there is no other noise from other applications. 
> 
> I used idleness profile (https://fio.readthedocs.io/en/latest/fio_doc.html#cpu-idleness-profiling) before it's basically taking all the remaining CPU with a lowest priority task, which means those are the CPU that can be used in the environment. So we can say for more if it's higher, the CPU is more `idle`.

yasker commented:
I don't see a quick way to replace the functionality of the idleness profiling at the moment. 

How about we make it an option, and adjust the output format accordingly? Better set to off by default then.

derekbit commented:
> I don't see a quick way to replace the functionality of the idleness profiling at the moment.
> 
> How about we make it an option, and adjust the output format accordingly? Better set to off by default then.

I agree with your proposal and will re-submit it.

derekbit committed:
Make CPU idleness profiling as an option

The CPU idleness introduces extra overhead and storage performance drop.
By default, the profiling is disabled.

Signed-off-by: Derek Su <derek.su@suse.com>
